utils/numcheck/calcderivs

```diff
@@ -157,73 +171,74 @@ def read_reference_results(autotest0, calcforces, calclatderivs, latvecs):
 
 
 def calculate_reference(binary, reffile, coords, specienames, species, origin,
-                        latvecs, chargeRestart):
+                        latvecs, charge_restart):
     """Calculates reference system"""
 
     writegen("geo.gen", (specienames, species, coords, origin, latvecs))
-    if (chargeRestart):
-        shutil.copy2 (chargeRestart.name, CHARGES_FILE)
+    if (charge_restart):
+        shutil.copy2(charge_restart.name, CHARGES_FILE)
     subprocess.call([binary])
     shutil.move("autotest.tag", reffile)
 
 
 def calculate_forces(binary, disp, coords, specienames, species, origin,
-                     latvecs, chargeRestart):
+                     latvecs, charge_restart, direction, atoms):
     """Calculates forces by finite differences"""
 
     cart =('x','y','z')
     delta =('-','+')
     energy = np.empty((2,), dtype=float)
-    forces = np.empty((len(coords), 3), dtype=float)
-    for iat in range(len(coords)):
-        for ii in range(3):
+    forces = np.empty((len(atoms), len(direction)), dtype=float)
+    for aa, iat in enumerate(atoms):
+        for ii, coord in enumerate(direction):
             for jj in range(2):
                 newcoords = np.array(coords)
-                newcoords[iat][ii] += float(2 * jj - 1) * disp
+                newcoords[iat][coord] += float(2 * jj - 1) * disp
                 writegen("geo.gen", (specienames, species, newcoords, origin,
                                      latvecs))
-                if (chargeRestart):
-                    shutil.copy2 (chargeRestart.name, CHARGES_FILE)
+                if (charge_restart):
+                    shutil.copy2(charge_restart.name, CHARGES_FILE)
                 subprocess.call([binary])
                 fp = open("autotest.tag", "r")
                 txt = fp.read()
                 fp.close()
                 match = ENERGY_PATTERN.search(txt)
-                print("iat: %2d, dir: %s, delta: %s" % (iat, cart[ii],delta[jj]))
+                print("iat: %2d, dir: %s, delta: %s" % (iat + 1, cart[coord], delta[jj]))
                 if match:
                     energy[jj] = float(match.group("value"))
                     print("energy:", energy[jj])
                 else:
                     raise ValueError("No energy match found in autotest.tag!")
-            forces[iat][ii] = (energy[0] - energy[1]) / (2.0 * disp)
+            forces[aa][ii] = (energy[0] - energy[1]) / (2.0 * disp)
     return forces
 
 
 def calculate_latderivs(binary, disp, coords, latvecs, specienames, species,
-                        origin, chargeRestart):
+                        origin, charge_restart, direction):
     """Calculates lattice derivatives by finite differences"""
 
+    cart =('x','y','z')
     delta =('-','+')
     energy = np.empty((2,), dtype=float)
-    latderivs = np.empty((3, 3), dtype=float)
-    for ii in range(3):
+    latderivs = np.empty((3, len(direction)), dtype=float)
+    for ii, coord in enumerate(direction):
```

Comment:
`ii` is not used inside the loop

Reply:
It is used at the end of that loop for filling into `latderivs`.

https://github.com/dftbplus/dftbplus/blob/d20a9e5d9e96274ac95134ee981dc0242058b349/utils/numcheck/calcderivs#L245-L248

Reply:
Sorry, I've missed that.